sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/cache/CacheFactory.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.spanner.changestreams.cache;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.io.gcp.spanner.changestreams.dao.DaoFactory;
+import org.joda.time.Duration;
+
+public class CacheFactory implements Serializable {
+
+  private static final long serialVersionUID = -8722905670370252723L;
+  private static final Map<Long, WatermarkCache> WATERMARK_CACHE = new ConcurrentHashMap<>();
+  private static final AtomicLong CACHE_ID = new AtomicLong();
+  private final DaoFactory daoFactory;
+  private final @Nullable Duration refreshRate;
+
+  public CacheFactory(DaoFactory daoFactory, @Nullable Duration watermarkRefreshRate) {
+    this.daoFactory = daoFactory;
+    this.refreshRate = watermarkRefreshRate;
+  }
+
+  public static long generateCacheId() {
```

Comment:
I wonder if this could be internal to the cache factory to simplify usage?


It seems that 
final long cacheId = CACHE_ID.getAndIncrement();
 coudl be member of the CacheFactory itself, so each factory has it's own id. Then the id is then serialized with it. And then when deserializing the factory (as part of serialized dofn) the multiple factories will have the same id and share the cache.

Reply:
Done